appdatabase/migrations/sql/1677681143_accounts_table_type_column_update.up.sql

```diff
@@ -0,0 +1 @@
+UPDATE accounts SET type = "seed" WHERE derived_from IN (SELECT derived_from FROM accounts WHERE type = "seed");
```

Comment:
this sql seems very very dangerous.. I ran it on my local machine, guess what happened? all rows affected.. and all of its type updated to "seed", before this, there were different types: seed/key/generated/""(empty), do you think it's normal?  🤔

Reply:
@qfrank thanks for the review.

Exactly that was my intention, to update `type` field for all accounts generated from the same seed phrase. Without that line we cannot be backward compatible.

> this sql seems very very dangerous.. 

I am wondering what's that dangerous, the only possible values for `type` column are `generated`, `seed`, `key`, `watch` and empty string (but the empty string is special case, refers only to chat address and default wallet address). 

Having that in mind if you read the query carefully you will see that only the `type` column of accounts which have `seed` as a value for the `type` column will be updated to `seed` value, what actually was my intention.

Reply:
> there were different types: seed/key/generated/""(empty)

I am not sure how it's possible that all your accounts are affected when we're checking if `derived_from` is in the set returned by this query `SELECT derived_from FROM accounts WHERE type = "seed"`

That makes me think that your `derived_from` column was empty for all your accounts before you run migration script, what further means that you're using a very very old account. Is that the case maybe?

Reply:
> all your accounts are affected 

that's what i mean dagnerous.
subquery SELECT derived_from FROM accounts WHERE type = "seed" will return set("","0x..") which contains empty string, that's why all my accounts were affected. we should take care of empty string, right? my accounts are not very very old ... 
<img width="1356" alt="image" src="https://user-images.githubusercontent.com/12406719/222353505-d848e6e9-4c6c-46a4-ae36-b39cfa8c304d.png">

Reply:
Seems either that your Status account is very old or mobile app somehow adds account without set `derived_from`.

> we should take care of empty string, right?

You must not have at all any account with `type` value set to `seed` or `generated` with an empty `derived_from`. Cause that is unallowed, an incorrect `accounts` table state. But seems that your `accounts` table had exactly that, an empty string set for `derived_from` column.


The only accounts from the `accounts` table which may have an empty `derived_from` value are:
- Status chat account, that's an account with `m/43'/60'/1581'/0'/0` derivation path
- accounts added via importing private key, those with value `key` for `type` column

Reply:
@saledjenic derived_from might not be used today by the mobile app as mobile don't need to derive new account yet

Reply:
@alaibe even they don't use it, they have to set it when user is creating new account via mobile app and then use the same account in the desktop app, cause otherwise keypairs on the desktop app side will be broken. Cause all accounts with the same `derived_from` value belong to the single keypair.

Reply:
@alaibe @qfrank but thinking now, I am not sure how we can set `derived_from` field for already created accounts?

We need to have it set for:
- Status default wallet account (path `m/44'/60'/0'/0/0`), profile keypair
- all other derivations from the profile keypair
- for account created by importing seed phrase, imported keypair
- for all derivations from the imported keypair

Reply:
> @alaibe even they don't use it, they have to set it when user is creating new account via mobile app and then use the same account in the desktop app, cause otherwise keypairs on the desktop app side will be broken. Cause all accounts with the same `derived_from` value belong to the single keypair.

my quick reply: even mobile don't use it, we should consider earlier user for status mobile, right? 🙂

Reply:
> even mobile don't use it, we should consider earlier user for status mobile, right? 

I am not sure when exactly in the past `derived_from` was introduce, but definitely from that point onwards adding accounts via mobile and desktop app divergates. All in all at the moment we need to find a way how to overcome that and as I see handling most likely should be on the mobile app side.
  
This is definitely an obstacle we have to discuss with @John-44

Reply:
add `derived_from != ""` seems safe to mobile db, but i want hear something from @cammellos 🙂

Reply:
@saledjenic @alaibe this needs to be addressed by whoever created the issue and introduced the breaking change, we can't have breaking changes not handled and let someone else having to deal with the consequences, this should be agreed before hand we make a breaking change.

The solution proposed are adding considerable development work and should have been discussed before making a breaking change. cc @John-44 

Is there any alternative solution? What's the difference between `derived_from` and `account_type`?

Reply:
@cammellos Hey, I wouldn't say it is a breaking change. As nothing is broken from both apps. It is only that in mobile wallet the derived from is not being used yet.

The derived from is to allow us to derive more than one account from a seed phrase. Which is not yet a feature implemented in mobile AFAIK.

Alternative solution would be create edge cases and limit the possible feature on such account. However that would create additional maintenance work for desktop and mobile and this would be an edge case we have to deal with

Reply:
> What's the difference between derived_from and account_type?

@cammellos `account_type` is there to differ accounts, is an account added by:
- deriving from a profile keypair (master key a default Status wallet account is derived from), in this case `account_type` is set to `generated`
- importing seed phrase or deriving from master key added by importing seed phrase, in this case `account_type` is set to `seed`
- importing private key, in this case `account_type` is set to `key`
- adding an account address, watch only account is added, in this case `account_type` is set to `watch` 

`derived_from` is there that we can know which accounts are derived from the same master key.

Reply:
@saledjenic @alaibe thanks for the explanation.

I checked the code on mobile, and it seems we do set the type to `seed` when creating from a seed phrase, so that should not really apply to us ( I haven't really run the code, just checked it), but maybe I am missing something, would you mind explaining why this change is necessary and when we created accounts with the wrong type?

Reply:
This will be needed by the new Mobile wallet designs as well, we should aim for the new mobile wallet to be functionally identical with the desktop wallet

Reply:
@cammellos trying to explain below.

Just to be on the same page, since mobile app hasn't introduced keypairs yet, let's define a "keypair" as term.
A single keypair is defined by:
- name, which needs to be unique across the app (not added yet on the desktop app side, will be in the next step) 
- accounts, one or more, but all of them have to be derived from the same master key

<ins>For example</ins>:
When user imports a seed phrase into the wallet, a new account `Account1` is generated from that seed phrase and its type is set to `seed`. This added `Account1` belongs to a new keypair named `MyKeypair1`. 

> it seems we do set the type to seed when creating from a seed phrase

What you have in mobile app is adding a single account from the seed phrase, setting the `type` to `seed` for that account. But now what we have in the desktop app, and I guess you don't have in mobile app, is an ability to let user creates another account for a keypair (the same master key) he already added earlier.

<ins>Continuation to the example above</ins>:
After adding `Account2` to the keypair `MyKeypair1` that keypair will have 2 accounts `Account1` and `Account2` and both of those accounts will have the same `derived_from` address and both of them will have the same `type` set to `seed`.

That's how it should be, but in the desktop app, `type` in this case was set to `generated`

> would you mind explaining why this change is necessary and when we created accounts with the wrong type?

This change is needed cause it fixes an issue on the desktop app, dating from who knows when, by updating `type` for all accounts generated from a keypair which is created by importing the seed phrase.

Reply:
> but in the desktop app, type in this case was set to generated (that's fixed by this PR).

Hi @saledjenic , could you explain why in this case it should not set to generated? according to your example, i think type for Account1 should be `seed`, type for Account2 should be `generated` since they have the same `derived_from` address or we can say because they have the same master key. that's how `generated` be defined from my understanding

Reply:
Are you ok with the changes I proposed in my previous comment, so I can do them here and overcome this in that way?

Reply:
I am currently working on updating add account flow in the desktop app (according to the latest UI changes for that) and within those changes I will add new column `keypair_name` column to `accounts` table (that's a new request).

Reply:
what about introducing new type? let's say it's `keypair-generated`, i feel the way you create account 0x22/0x23 is not the same as the old way create `generated`, but you didn't explain it in your reply

address | type | name | derived_from
-- | -- | -- | --
0x11 |   | Status account | 0x1111
0x12 | generated | Derived1 | empty
0x13 | generated | Derived2 | empty
0x21 | seed | SImported1 | 0x2222
0x22 | keypair-generated | SImpoGen11 | 0x2222
0x23 | keypair-generated | SImpoGen12 | 0x2222

Reply:
> Are you ok with the changes I proposed in my previous comment, so I can do them here and overcome this in that way?

It's okay if @cammellos says it's okay, i think my okay doesn't count in this situation 😅

Reply:
> @cammellos trying to explain below.
> 
> Just to be on the same page, since mobile app hasn't introduced keypairs yet, let's define a "keypair" as a term. A single keypair is defined by:
> 
> * **name**, which needs to be unique across the app (not added yet on the desktop app side, will be in the next step)
> * **accounts**, one or more, but all of them have to be derived from the same master key
> 
> For example: When user imports a seed phrase into the wallet, a new account `Account1` is generated from that seed phrase and its type is set to `seed`. This added `Account1` belongs to a new keypair named `MyKeypair1`.
> 
> > it seems we do set the type to seed when creating from a seed phrase
> 
> What you have in mobile app is adding a single account from the seed phrase, setting the `type` to `seed` for that account, that's ok, we have that feature as well. But now what we have in the desktop app, and I guess you don't have in mobile app, is an ability to let user creates another account for a keypair (the same master key) he already added earlier.
> 
> Continuation to the example above: After adding `Account2` to the keypair `MyKeypair1` that keypair will have 2 accounts `Account1` and `Account2` and both of those accounts will have the same `derived_from` address and both of them will have the same `type` set to `seed`.
> 
> That's how it should be, but in the desktop app, `type` in this case was set to `generated` (that's fixed by this PR).
> 
> > would you mind explaining why this change is necessary and when we created accounts with the wrong type?
> 
> This change is needed cause it fixes an issue on the desktop app, dating from who knows when, by updating `type` for all accounts generated from a keypair which is created by importing the seed phrase.

Thanks for the explanation
with regard to that, no migration I believe is needed on mobile, since we have not allowed users to derive a wallet from an imported seed yet, so there won't be any account that has the bug you mentioned (which is only present on desktop, is that correct?)
If so, the migration after the changes asked by @qfrank should work fine (i.e as long as we are migrating account that have a `derived_from` set, it should not impact existing mobile users, and their database is consistent since they don't have any of these new type of accounts as we never allowed them to generate any)

Is that a correct representation of the state of things or I am missing something?

Reply:
> what about introducing new type? let's say it's keypair-generated

I don't see need for new type. Why would we introduce a new type if accounts `0x21`, `0x22` and `0x23` are functionally wise the same?

>  i feel the way you create account 0x22/0x23 is not the same as the old way create generated

Not sure what old way you're referring to? Before my changes done here we were using [this function](https://github.com/status-im/status-go/blob/1b24342a75d99b8e2601d7639295899ea75b6fb3/services/accounts/accounts.go#L290) for adding account generated from already imported seed phrase and it sets the `type` column to `seed`. In this PR we're still using the same call, but I change the `generateAccountPasswordVerified` that it sets `type` correctly now.

Reply:
@cammellos yes, that's correct, after changes I am going to add no migration will be needed for mobile users.

Will post here and @qfrank may give another look and test if mobile app users are somehow affected by the changes.

Reply:
note: mobile won't take care of field `derived_from`(so it will be empty) when import account from seed phrase

Reply:
> My proposal was (which will work for you without any drawbacks):

to update the query with derived_from != "" to look like this SELECT derived_from FROM accounts WHERE type = "seed" AND derived_from != ""
ignore accounts with empty derived_from from syncing

3. what if we only apply the second proposal

Reply:
1 - Addresses 0x21, 0x22 and 0x23 should be the same? they are not the same in local mobile test
How did u create addresses for this case in mobile, I thought u didn't support yet generating more than one account from a seed phrase?

2- If i understand correctly we are trying to limit the scope of the value "generated" to only being "generated" as they were before no?

Reply:
@qfrank I've just committed the the update, please check if this PR anyhow affects the mobile app users?

It contains `derived_from != ""` for the query and added a comment to the `HandleSyncWalletAccount` function where we need to do some handling later (will do that in another PR when the time comes, once I finished my current work and get back to syncing).

Reply:
> 1 - Addresses 0x21, 0x22 and 0x23 should be the same? they are not the same in local mobile test How did u create addresses for this case in mobile, I thought u didn't support yet generating more than one account from a seed phrase?
> 
> 2- If i understand correctly we are trying to limit the scope of the value "generated" to only being "generated" as they were before no?

Thank you for your reply, make sense @alaibe

Reply:
> what if we only apply the second proposal? i mean we don't need the migration appdatabase/migrations/sql/1677681143_accounts_table_type_column_update.up.sql

- accounts `0x21`, `0x22` and `0x23` are functionally wise the same, no point in having some of them with `type` set to `seed` and some of them to `generated`. An account added along with importing a seed phrase or new accounts generated later from already imported seed phrase should have `type` set to `seed`, it's more intuitive and cleaner that way. 
- the desktop app already uses `type` in several places, meaning that without updating `type` to `seed` for accounts generated from imported seed phrase we will have some keypairs broken if the account created in the moment of importing a seed phrase gets deleted (since only it has `seed` for the `type`).